support-models/src/main/scala/com/gu/support/workers/states/SendThankYouEmailState.scala

```diff
@@ -1,37 +1,100 @@
 package com.gu.support.workers.states
 
-import java.util.UUID
-
-import com.gu.support.encoding.CustomCodecs.{decodeLocalTime, encodeLocalTime}
+import com.gu.support.encoding.DiscriminatedType
+import com.gu.support.encoding.CustomCodecs._
 import com.gu.support.promotions.PromoCode
-import com.gu.support.redemptions.RedemptionData
+import com.gu.support.workers.GiftRecipient.{DigitalSubscriptionGiftRecipient, WeeklyGiftRecipient}
 import com.gu.support.workers.{PaymentMethod, SalesforceContactRecord, User, _}
 import org.joda.time.LocalDate
 
-case class SendThankYouEmailState(
-  requestId: UUID,
-  user: User,
-  giftPurchase: Option[GiftPurchase],
-  product: ProductType,
-  analyticsInfo: AnalyticsInfo,
-  paymentOrRedemptionData: Either[PaymentMethodWithSchedule, RedemptionData],
-  firstDeliveryDate: Option[LocalDate],
-  promoCode: Option[PromoCode],
-  salesForceContact: SalesforceContactRecord,
-  accountNumber: String,
-  subscriptionNumber: String,
-  acquisitionData: Option[AcquisitionData]
-) extends SendAcquisitionEventState
-
-import com.gu.support.encoding.Codec
-import com.gu.support.encoding.Codec._
+sealed trait SendThankYouEmailState extends StepFunctionUserState {
+  def product: ProductType
+}
 
 object SendThankYouEmailState {
-  import com.gu.support.encoding.CustomCodecs._
-  implicit val codec: Codec[SendThankYouEmailState] = deriveCodec
-}
 
-case class PaymentMethodWithSchedule(paymentMethod: PaymentMethod, paymentSchedule: PaymentSchedule)
-object PaymentMethodWithSchedule {
-  implicit val codec: Codec[PaymentMethodWithSchedule] = deriveCodec
+  sealed trait SendThankYouEmailDigitalSubscriptionState extends SendThankYouEmailState {
+    override def product: DigitalPack
+  }
+
+  case class SendThankYouEmailContributionState(
+    user: User,
+    salesForceContact: SalesforceContactRecord,
+    product: Contribution,
+    paymentMethod: PaymentMethod,
+    accountNumber: String,
+  ) extends SendThankYouEmailState
+
+  case class SendThankYouEmailDigitalSubscriptionDirectPurchaseState(
+    user: User,
+    salesForceContact: SalesforceContactRecord,
+    product: DigitalPack,
+    paymentMethod: PaymentMethod,
+    paymentSchedule: PaymentSchedule,
+    promoCode: Option[PromoCode],
+    accountNumber: String,
+    subscriptionNumber: String,
+  ) extends SendThankYouEmailDigitalSubscriptionState
+
+  case class SendThankYouEmailDigitalSubscriptionGiftPurchaseState(
+    user: User,
+    salesForceContact: SalesforceContactRecord,
+    product: DigitalPack,
+    giftRecipient: DigitalSubscriptionGiftRecipient,
+    giftCode: GeneratedGiftCode,
+    lastRedemptionDate: LocalDate,
+    paymentMethod: PaymentMethod,
+    paymentSchedule: PaymentSchedule,
+    accountNumber: String,
+  ) extends SendThankYouEmailDigitalSubscriptionState
```

Comment:
following the earlier chat, we may need `promoCode: Option[PromoCode],` in here too

Reply:
added this in